File: ios/Classes/Bridges/PurchaseResultBridge.swift
import Flutter
import SuperwallKit

public class PurchaseResultBridge: BridgeInstance {
  var purchaseResult: PurchaseResult {
    assertionFailure("Subclasses must implement")
    return .failed(PurchaseResultError(message: "Subclasses must implement purchaseResult"))
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    result(FlutterMethodNotImplemented)
  }
}

public class PurchaseResultCancelledBridge: PurchaseResultBridge {
  class override var bridgeClass: BridgeClass { "PurchaseResultCancelledBridge" }
  override var purchaseResult: PurchaseResult { return .cancelled }
}

public class PurchaseResultPurchasedBridge: PurchaseResultBridge {
  class override var bridgeClass: BridgeClass { "PurchaseResultPurchasedBridge" }
  override var purchaseResult: PurchaseResult { return .purchased }
}

public class PurchaseResultRestoredBridge: PurchaseResultBridge {
  class override var bridgeClass: BridgeClass { "PurchaseResultRestoredBridge" }
  override var purchaseResult: PurchaseResult { return .restored }
}

public class PurchaseResultPendingBridge: PurchaseResultBridge {
  class override var bridgeClass: BridgeClass { "PurchaseResultPendingBridge" }
  override var purchaseResult: PurchaseResult { return .pending }
}

public class PurchaseResultFailedBridge: PurchaseResultBridge {
  class override var bridgeClass: BridgeClass { "PurchaseResultFailedBridge" }
  override var purchaseResult: PurchaseResult { .failed(error) }

  let error: Error

  required init(bridgeId: BridgeId, initializationArgs: [String : Any]? = nil) {
    guard let error = initializationArgs?["error"] as? String else {
      fatalError("Attempting to create `PurchaseResultFailedBridge` without providing `error`.")
    }

    self.error = PurchaseResultError(message: error)
    super.init(bridgeId: bridgeId, initializationArgs: initializationArgs)
  }
}

struct PurchaseResultError: Error {
  let message: String
}

File: android/src/main/kotlin/com/superwall/superwallkit_flutter/bridges/PurchaseResultBridge.kt
package com.superwall.superwallkit_flutter.bridges

import android.content.Context
import com.superwall.sdk.delegate.PurchaseResult
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

abstract class PurchaseResultBridge(
    context: Context,
    bridgeId: BridgeId,
    initializationArgs: Map<String, Any>? = null
) : BridgeInstance(context, bridgeId, initializationArgs), MethodChannel.MethodCallHandler {

    abstract val purchaseResult: PurchaseResult

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        result.notImplemented()
    }
}

class PurchaseResultCancelledBridge(
    context: Context,
    bridgeId: BridgeId,
    initializationArgs: Map<String, Any>? = null
) : PurchaseResultBridge(context, bridgeId, initializationArgs) {
    companion object { fun bridgeClass(): BridgeClass = "PurchaseResultCancelledBridge" }
    override val purchaseResult: PurchaseResult = PurchaseResult.Cancelled()
}

class PurchaseResultPurchasedBridge(
    context: Context,
    bridgeId: BridgeId,
    initializationArgs: Map<String, Any>? = null
) : PurchaseResultBridge(context, bridgeId, initializationArgs) {
    companion object { fun bridgeClass(): BridgeClass = "PurchaseResultPurchasedBridge" }
    override val purchaseResult: PurchaseResult = PurchaseResult.Purchased()
}

class PurchaseResultPendingBridge(
    context: Context,
    bridgeId: BridgeId,
    initializationArgs: Map<String, Any>? = null
) : PurchaseResultBridge(context, bridgeId, initializationArgs) {
    companion object { fun bridgeClass(): BridgeClass = "PurchaseResultPendingBridge" }
    override val purchaseResult: PurchaseResult = PurchaseResult.Pending()
}

class PurchaseResultFailedBridge(
    context: Context,
    bridgeId: BridgeId,
    initializationArgs: Map<String, Any>? = null
) : PurchaseResultBridge(context, bridgeId, initializationArgs) {
    companion object { fun bridgeClass(): BridgeClass = "PurchaseResultFailedBridge" }
    override val purchaseResult: PurchaseResult

    init {
        val errorString = initializationArgs?.get("error") as? String
            ?: throw IllegalArgumentException("Attempting to create `PurchaseResultFailedBridge` without providing `error`.")
        purchaseResult = PurchaseResult.Failed(errorString)
    }
}